### PR TITLE
Initial commit

### PR DESCRIPTION
**Pull request by badkk**
Initial AutoRestRef Tool

### CODE REVIEW

File: AutoRestRef/ARRAgent.cs [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+using HtmlAgilityPack;
+using Newtonsoft.Json;
+using AutoRestRef.Templates;
+using AutoRestRef.DataAccess;
+
+namespace AutoRestRef
+{
+    class ARRAgent
+    { 
+
```
Comment (kimizhu) [inline AutoRestRef/ARRAgent.cs:14]: empty line

File: AutoRestRef/ARRAgent.cs [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+            var content = RemoteFileAccess.GetOnlineContent(serviceUrl);
+            const string defaultDesContent = "";
+            if (content != null)
+            {
+                //parse the html
+                Console.WriteLine($"Parse description from {serviceUrl}");
+                var doc = new HtmlDocument();
+                doc.LoadHtml(content);
+                var desNodes = doc.DocumentNode.SelectNodes("//main/p");
+                var value = desNodes == null ? defaultDesContent : desNodes.First().InnerText;
```
Comment (kimizhu) [inline AutoRestRef/ARRAgent.cs:26]: return desNodes == null ? defaultDesContent : desNodes.First().InnerText;

File: AutoRestRef/ARRAgent.cs [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+            {
+                //parse the html
+                Console.WriteLine($"Parse description from {serviceUrl}");
+                var doc = new HtmlDocument();
+                doc.LoadHtml(content);
+                var desNodes = doc.DocumentNode.SelectNodes("//main/p");
+                var value = desNodes == null ? defaultDesContent : desNodes.First().InnerText;           
+                return value;
+            }
+            else
```
Comment (kimizhu) [inline AutoRestRef/ARRAgent.cs:29]: put it before if 
if (content == null) return "";

File: AutoRestRef/ARRAgent.cs [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+            {
+                //if 404 or parse error then return ""
+                return defaultDesContent;
+            }
+        }
+        private static List<ServiceTemplate> GetServices(string tocUrl)
```
Comment (kimizhu) [inline AutoRestRef/ARRAgent.cs:35]: can we return IEnumerable instead?

File: AutoRestRef/ARRAgent.cs [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+using Newtonsoft.Json.Linq;
+using System.Linq;
+using HtmlAgilityPack;
+using Newtonsoft.Json;
+using AutoRestRef.Templates;
+using AutoRestRef.DataAccess;
+
+namespace AutoRestRef
+{
+    class ARRAgent
```
Comment (kimizhu) [inline AutoRestRef/ARRAgent.cs:12]: what does ARR mean?

File: AutoRestRef/ARRAgent.cs (hunk 2 above)
Comment (kimizhu) [reply AutoRestRef/ARRAgent.cs:26]: also use null-conditional-operators
https://docs.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-6#null-conditional-operators

File: AutoRestRef/ARRAgent.cs [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+                    return new OutputTemplate(name, scope, services);
+                }).ToList();
+                return rsl;
+            }
+            else
+            { 
+                return null;
+            }
+        }
+        public static void Run(string inputFilePath, string outputFilePath)
```
Comment (kimizhu) [inline AutoRestRef/ARRAgent.cs:78]: better to put public method before private method. more readable.

File: AutoRestRef/ARRAgent.cs [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+            {
+                var restApiJArray = JArray.Parse(content);
+                var rsl = restApiJArray.Select(obj =>
+                {
+                    var name = (string)obj["name"];
+                    var scope = (string)obj["scope"];
+                    var services = GetServices((string)obj["toc_url"]);
+                    return new OutputTemplate(name, scope, services);
+                }).ToList();
+                return rsl;
```
Comment (kimizhu) [inline AutoRestRef/ARRAgent.cs:71]: don't name this way. give them meaningful name

File: AutoRestRef/DataAccess/LocalFileAccess.cs [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+                return null;
+            }
+        }
+
+        public static bool WriteFile(string path, string text)
+        {
+            try
+            {
+                using (var sw = new StreamWriter(path))
+                    sw.WriteLine(text);
```
Comment (kimizhu) [inline AutoRestRef/DataAccess/LocalFileAccess.cs:28]: support unicode?

File: AutoRestRef/DataAccess/RemoteFileAccess.cs [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+{
+    class RemoteFileAccess
+    {
+        public static string GetOnlineContent(string url)
+        {
+            try
+            {
+                var client = new WebClient();
+                var uri = new Uri(url);
+                return Encoding.ASCII.GetString(client.DownloadData(uri));
```
Comment (kimizhu) [inline AutoRestRef/DataAccess/RemoteFileAccess.cs:15]: will this be blocked by anti crawler?

File: AutoRestRef/ARRAgent.cs (hunk 1 above)
Comment (badkk) [reply AutoRestRef/ARRAgent.cs:14]: Solved

File: AutoRestRef/ARRAgent.cs (hunk 2 above)
Comment (badkk) [reply AutoRestRef/ARRAgent.cs:26]: Solved

File: AutoRestRef/ARRAgent.cs (hunk 3 above)
Comment (badkk) [reply AutoRestRef/ARRAgent.cs:29]: Solved

File: AutoRestRef/ARRAgent.cs (hunk 5 above)
Comment (badkk) [reply AutoRestRef/ARRAgent.cs:12]: ARR meas AutoRestRef [haha]

File: AutoRestRef/ARRAgent.cs (hunk 6 above)
Comment (badkk) [reply AutoRestRef/ARRAgent.cs:78]: Solved

File: AutoRestRef/ARRAgent.cs (hunk 7 above)
Comment (badkk) [reply AutoRestRef/ARRAgent.cs:71]: Solved

File: AutoRestRef/DataAccess/LocalFileAccess.cs (hunk 8 above)
Comment (badkk) [reply AutoRestRef/DataAccess/LocalFileAccess.cs:28]: yes, it does

File: AutoRestRef/DataAccess/RemoteFileAccess.cs (hunk 9 above)
Comment (badkk) [reply AutoRestRef/DataAccess/RemoteFileAccess.cs:15]: Seems like this did not happen yet.

File: AutoRestRef/ARRAgent.cs (hunk 5 above)
Comment (kimizhu) [reply AutoRestRef/ARRAgent.cs:12]: so name it to AutoRestRefAgent？

**Pull request merged**